python/ray/rllib/agents/ddpg/ddpg_policy_graph.py

```diff
@@ -343,39 +198,43 @@ def __init__(self, observation_space, action_space, config):
         q_batchnorm_update_ops = list(
             set(tf.get_collection(tf.GraphKeys.UPDATE_OPS)) - prev_update_ops)
 
-        # target q network evalution
+        # target q network evaluation
         with tf.variable_scope(Q_TARGET_SCOPE) as scope:
             q_tp1, _ = self._build_q_network(self.obs_tp1, observation_space,
-                                             action_space,
-                                             output_actions_estimated)
+                                             action_space, policy_tp1_smoothed)
             target_q_func_vars = _scope_vars(scope.name)
         if self.config["twin_q"]:
             with tf.variable_scope(TWIN_Q_TARGET_SCOPE) as scope:
-                twin_q_tp1, _ = self._build_q_network(
-                    self.obs_tp1, observation_space, action_space,
-                    output_actions_estimated)
+                twin_q_tp1, _ = self._build_q_network(self.obs_tp1,
+                                                      observation_space,
+                                                      action_space,
+                                                      policy_tp1_smoothed)
                 twin_target_q_func_vars = _scope_vars(scope.name)
 
         if self.config["twin_q"]:
-            self.loss = self._build_actor_critic_loss(
-                q_t, q_tp1, q_tp0, twin_q_t=twin_q_t, twin_q_tp1=twin_q_tp1)
+            self.critic_loss, self.actor_loss, self.td_error \
+                = self._build_actor_critic_loss(
+                    q_t, q_tp1, q_t_det_policy, twin_q_t=twin_q_t,
+                    twin_q_tp1=twin_q_tp1)
         else:
-            self.loss = self._build_actor_critic_loss(q_t, q_tp1, q_tp0)
+            self.critic_loss, self.actor_loss, self.td_error \
+                = self._build_actor_critic_loss(
+                    q_t, q_tp1, q_t_det_policy)
 
         if config["l2_reg"] is not None:
-            for var in self.p_func_vars:
+            for var in self.policy_vars:
                 if "bias" not in var.name:
-                    self.loss.actor_loss += (
-                        config["l2_reg"] * 0.5 * tf.nn.l2_loss(var))
+                    self.actor_loss += (config["l2_reg"] * 0.5 *
+                                        tf.nn.l2_loss(var))
             for var in self.q_func_vars:
                 if "bias" not in var.name:
-                    self.loss.critic_loss += (
-                        config["l2_reg"] * 0.5 * tf.nn.l2_loss(var))
+                    self.critic_loss += (config["l2_reg"] * 0.5 *
+                                         tf.nn.l2_loss(var))
             if self.config["twin_q"]:
                 for var in self.twin_q_func_vars:
```

Comment:
where is the definition (I mean as left value) of `self.twin_q_func_vars`.

Reply:
On line 197, where it builds the second Q-network (`self.twin_q_func_vars = _scope_vars(scope.name)`).

Reply:
(also, I added a small commit fixing the other two issues you identified)